### ISSUE TITLE
Support for Gradle composite builds

### ISSUE THREAD

**Issue by zharinov** [renovatebot/renovate#9410]
Duplicate of #8692

**Comment by zharinov** [comment]
Example files:
- https://github.com/donbeave/graphql-java-datetime
  - [buildSrc/src/main/java/Versions.kt](https://github.com/donbeave/graphql-java-datetime/blob/5417ee4de4acb3d2fa432568b89618ecac6e8366/buildSrc/src/main/java/Versions.kt)
  - [build.gradle.kts](https://github.com/donbeave/graphql-java-datetime/blob/5417ee4de4acb3d2fa432568b89618ecac6e8366/build.gradle.kts)

More:
- [`radikushch/GAlertButton:`dependencies/src/main/java/com/example/dependencies/GlobalVersions.kt](https://github.com/radikushch/GAlertButton/blob/4a6328c365a041f27c35e3094fd7a3779573ec58/dependencies/src/main/java/com/example/dependencies/GlobalVersions.kt)
- [`square/okhttp:`buildSrc/src/main/kotlin/deps.kt](https://github.com/square/okhttp/blob/f8fd4d08decf697013008b05ad7d2be10a648358/buildSrc/src/main/kotlin/deps.kt)
- [`PokerAlpaka/kunde:`buildSrc/src/main/kotlin/Versions.kt](https://github.com/PokerAlpaka/kunde/blob/c0b7c83e36091da3246dfb1e83c1a5426c9e5b8e/buildSrc/src/main/kotlin/Versions.kt)
- [`thewca/tnoodle:`buildSrc/src/main/kotlin/dependencies/Versions.kt](https://github.com/thewca/tnoodle/blob/125a511415785203527a19f72a95ef27b930b165/buildSrc/src/main/kotlin/dependencies/Versions.kt)

**Comment by rarkins** [comment]
Do we need to discover these files using `fileMatch` or will the `build.gradle` reference them?

**Comment by rarkins** [comment]
Also, is this the same as #8692 ?

**Comment by zharinov** [comment]
Oh, I forgot about it. Yes, it is.

**Comment by rarkins** [comment]
Maybe close this as a duplicate and copy over the links from here

**Comment by HonkingGoose** [comment]
Duplicate of #8692

**Comment by zharinov** [comment]
Desired functionality turned to be a part of another feature

**Comment by zharinov** [comment]
Probably should be higher priority than #8692

**Comment by Churro** [comment]
PR #21892 introduces parsing support for Kotlin source files with object declarations like `object X { ... }` and single variable assignments, such as `val dep = "..."` 

As of now, renovate doesn't recognize composite plugins explicitly and if you place e.g. a `Deps.kt` into a composite plugin outside of `buildSrc`, renovate wouldn't capture it. However, with the parsing support added with #21892, `.kt` files of composite plugins can be parsed in an opt-in fashion. 

E.g., the following renovate config would match a file like `includedBuild/dependencies/src/main/kotlin/com/badoo/reaktive/dependencies/Deps.kt` (ex. [here](https://github.com/badoo/Reaktive/blob/master/includedBuild/dependencies/src/main/kotlin/com/badoo/reaktive/dependencies/Deps.kt)):

  ```json
  {
    "gradle": {
      "fileMatch": ["^includedBuild/.+\\.kt$"]
    }
  }
  ```

**Comment by mlschechter** [comment]
It looks like all the functionality requested here is addressed via other released modifications - is that correct?

**Comment by Churro** [comment]
> It looks like all the functionality requested here is addressed via other released modifications - is that correct?

Yes and no. Everything that's mentioned in the examples above can already be matched by renovate. Matching `.kt` files works as described [here](https://github.com/renovatebot/renovate/issues/9410#issuecomment-1528830291) (probably it would make sense to add this to the public docs btw).

What's currently not supported are actual composite builds, as illustrated [here](https://docs.gradle.org/current/userguide/composite_builds.html#defining_composite_builds). So if you have a project that declares a binary dependency on a library that is included from a composite build, e.g., `some:internal:1.0.0`, you'll see a warning in the dependency dashboard if this dep is not yet present in one of the listed registries. In other words, there is no detection of composite builds per se.

To add real support for composite builds, it might be necessary to parse `includeBuild` directives from gradle settings. However, this info alone would be too imprecise to match composite builds, as you also need to know the current group and version, which are often only resolved at build time.
